### DISCUSSION TITLE
About memory aliasing

### DISCUSSION THREAD

**Discussion by ghost** [axboe/liburing#938] [Q&A]
Is the `io_uring` structure unaliased? In other words, if I allocate this structure, is my pointer the only pointer that points to this allocated memory?
This is important because Rust's mutable references require them to be unaliased, otherwise UnsafeCell must be used.

**Comment by ghost** [comment]
https://faultlore.com/blah/fix-rust-pointers/#aliasing
> Unaliased memory cannot be “randomly” modified by something seemingly “unrelated” (we’ll get to what that means in the next section).

**Comment by ghost** [comment]
I'm currently using UnsafeCell, which may be unsound, but mutable references to the io_uring struct are almost certainly unsound.

**Comment by ghost** [comment]
That's why I prefer using handles instead of structs. Handles are immutable because they are simply numbers, whereas structs are mutable and require consideration of aliasing and other issues.

**Comment by cmazakas** [comment]
In general, no, I wouldn't assume that you can safely apply `restrict` to the `struct io_uring*`.

For your Rust code, you should just stick with using the raw pointer, there's no need for `UnsafeCell`, imo.

You use `UnsafeCell` when you need to mutate a single data structure through various `&self` member functions, which isn't really how the liburing API works at all.

Looking at how I did it in Rust last time, I had a helper C function:
```c
struct io_uring *rio_setup(unsigned int entries, unsigned int flags)
{
  struct io_uring *ring = malloc(sizeof(*ring));
  int errnum = io_uring_queue_init(entries, ring, flags);
  if (0 != errnum)
  {
    return NULL;
  }
  return ring;
}
```
.

Re-examining this code, I'm actually not 100% sure this is correct but it worked fine for me with valgrind.

I'd recommend just doing this from Rust as well. The reason for this is because you avoid needing the definition of the `struct io_uring` from Rust which is quite gory.

**Reply by ghost** [reply]
I don't think bindgen is as bad as you think. You can take a look at the https://docs.rs/miniring and https://docs.rs/rawring bindings that I generated using rustbindgen.

**Reply by ghost** [reply]
The rawring improves bindgen result by making slight modifications to the io_uring header. This is always better than manually binding or using u64 instead of pointers.

**Reply by cmazakas** [reply]
Always better, huh?

I'd just keep a lot of those definitions hidden and just use the liburing-ffi lib.
